pkg/flowaggregator/certificate.go

```diff
@@ -0,0 +1,193 @@
+	maxAge    = time.Hour * 24 * 365       // one year self-signed certs
+)
+
+func generateCACertKey() (*x509.Certificate, *rsa.PrivateKey, []byte, error) {
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(2020),
+		Subject: pkix.Name{
+			CommonName: fmt.Sprintf("flow-aggregator-ca@%d", time.Now().Unix()),
+		},
+		NotBefore:             validFrom,
+		NotAfter:              validFrom.Add(maxAge),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	// generate private key for CA
+	caKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	// generate CA certificate
+	caCert, err := x509.CreateCertificate(rand.Reader, cert, cert, &caKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	caPEM := new(bytes.Buffer)
+	pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caCert,
+	})
+
+	return cert, caKey, caPEM.Bytes(), err
+}
+
+func generateCertKey(caCert *x509.Certificate, caKey *rsa.PrivateKey, isServer bool, flowAggregatorDNSName string) ([]byte, []byte, error) {
+	var cert *x509.Certificate
+	if isServer {
+		cert = &x509.Certificate{
+			SerialNumber: big.NewInt(2020),
+			Subject: pkix.Name{
+				CommonName: fmt.Sprintf("flow-aggregator-certificate@%d", time.Now().Unix()),
+			},
+			DNSNames:    []string{flowAggregatorDNSName},
+			NotBefore:   validFrom,
+			NotAfter:    validFrom.Add(maxAge),
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			KeyUsage:    x509.KeyUsageDigitalSignature,
+		}
+	} else {
+		cert = &x509.Certificate{
+			SerialNumber: big.NewInt(2020),
+			Subject: pkix.Name{
+				CommonName: fmt.Sprintf("flow-aggregator-certificate@%d", time.Now().Unix()),
+			},
+			NotBefore:   validFrom,
+			NotAfter:    validFrom.Add(maxAge),
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			KeyUsage:    x509.KeyUsageDigitalSignature,
+		}
+	}
+	// generate private key for certificate
+	certKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, err
+	}
+	// sign the certificate using CA certificate and key
+	certBytes, err := x509.CreateCertificate(rand.Reader, cert, caCert, &certKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPEM := new(bytes.Buffer)
+	pem.Encode(certPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+
+	certKeyPEM := new(bytes.Buffer)
+	pem.Encode(certKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(certKey),
+	})
+
+	return certPEM.Bytes(), certKeyPEM.Bytes(), nil
+}
+
+func syncCAAndClientCert(caCert, clientCert, clientKey []byte) error {
+	klog.Info("Syncing CA certificate with ConfigMap")
+
+	config, err := rest.InClusterConfig()
```

Comment:
I think I mentioned it before  but I think the K8s client should come from `cmd/flow-aggregator/`

Reply:
Updated. Thanks!